### DISCUSSION TITLE
LEFT JOINs: What is loaded into memory and what stays no disk?

### DISCUSSION THREAD

**Discussion by JonathanAquino-NextRoll** [ClickHouse/ClickHouse#45257] [Q&A] [answered]
I was just curious, if you do

    A LEFT JOIN B LEFT JOIN C LEFT JOIN D

Does A stay on disk while B, C, and D are loaded into memory? So the order of B, C, and D doesn't matter?

**Comment by mdonkers** [comment]
As far as I know, everything needs to be loaded into memory because its the only way to build the result set.
However ClickHouse would only read specific parts and columns that are needed for the result. E.g. only those columns needed for the SELECT result or for the JOIN itself. And load those parts from disk that would match the WHERE and/or JOIN clause.

And that is where the order might matter, but it depends on the full query and how ClickHouse would optimize it.
What you could do for example is set `set send_logs_level='trace'` , then execute the query and see how the query is optimized e.g. with PREWHERE clauses.

Hope that helps answering your question?

**Reply by JonathanAquino-NextRoll** [reply]
@mdonkers The reason I was saying ClickHouse loads the right side into memory is this line from https://clickhouse.com/docs/en/sql-reference/statements/select/join/

> By default, ClickHouse uses the [hash join](https://en.wikipedia.org/wiki/Hash_join) algorithm. ClickHouse takes the right_table and creates a hash table for it in RAM. 

So the right side should ideally be smaller than the left side, correct? So in my example, does the order of B, C, and D (which are all on the right side) matter?

**Reply by mdonkers** [reply]
[This section on memory limitations](https://clickhouse.com/docs/en/sql-reference/statements/select/join/#memory-limitations) is not about the table itself, but about the tuples that make up the JOIN (and then only those columns needed for the JOIN). It depends more on the number of records that will be selected for the JOIN, than the size of the actual table. But yes, if less records match on the right-side, than the hash table would be smaller.

Note though that ClickHouse would otherwise fall-back to a merge-join and that the size of the hash table is configurable.

**Reply by JonathanAquino-NextRoll** [reply]
@mdonkers ok so given that, does the order of B, C, and D matter for memory consumption (hash table size)?

I think it would not matter because table A is being left joined to each of them.

**Reply by UnamedRus** [reply, marked as answer]
> So the right side should ideally be smaller than the left side, correct? So in my example, does the order of B, C, and D (which are all on the right side) matter?

Yes, ClickHouse currently can't reorder join by itself (it's planned)
